### PR TITLE
fix: two A2/A3 defects behind the tile.select xfails (TSELS scratch reuse, int32 ordering compare)

### PR DESCRIPTION
**Pull request by lyfne123**
## What

`tile.sels` declares its `tmp` operand as an input, but on A2/A3 the intrinsic **writes** it. `pto.tsels` opens with a one-element store into `tmp` at offset 0, issued on the **scalar** pipe:

```cpp
// pto-isa a2a3/TSels.hpp:33-40
__ubuf__ SrcTile::DType* scalarPtr = (__ubuf__ SrcTile::DType*)__cce_get_tile_ptr(tmp);

*scalarPtr = scalar;                    // scalar pipe, one element at offset 0
PtoSetWaitFlag<PIPE_S, PIPE_V>();       // orders only this op's own RAW
set_cmpmask(scalarPtr);
pipe_barrier(PIPE_V);
for (unsigned i = 0; i < validRow; i++)
    vsel(..., SELMODE::VSEL_TENSOR_SCALAR_MODE);
```

Every other tile operand is read on the vector pipe, which makes TSELS the one op whose write originates on a different pipe than the reads it can race. The wait only runs *forward* (store → `set_cmpmask`); nothing retires an earlier **vector** reader of that buffer first.

The write never reaches the emitted `.pto` as a write — `MakeNaryCodegenPTO` passes operands positionally, carrying no read/write semantics — so no downstream sync insertion can see the WAR. MemoryReuse, which made the placement, has to prevent it.

## The failure

A clamp built from two `pl.tile.select`s corrupts exactly **element 0** of its compare source:

```python
t   = pl.load(v, [0, 0], [16, 128])
hi  = pl.tile.select(pl.tile.cmps(t,  2.0, cmp_type=3), t,   2.0)
out = pl.tile.select(pl.tile.cmps(t, -2.0, cmp_type=5), hi, -2.0)
```

```text
Mismatched elements: 1/2048
    [0] actual=-4.0, expected=-2.0
```

The allocation shows why — the second select's scratch inherited `t`'s buffer:

```text
t__ssa_v0          mem_vec_2  @0   sz=8192   pl.tile.load     <- read by BOTH tcmps
t__sel_tmp_tmp_v1  mem_vec_2  @0   sz=512    pl.tile.create   <- TSELS scratch
t__tmp_v2          mem_vec_2  @0   sz=8192   pl.tile.sels
```

`-4.0` is not noise. With `t[0] = -4.0`, the scalar store poisons it to `-2.0` before the second `tile.cmps` reads it, so `-2.0 >= -2.0` flips the mask to true and the select returns `hi[0] = -4.0` instead of the scalar `-2.0`.

## The fix

`tile.sels` never declared the write. The registry already has the vocabulary for it — `set_workspace_arg(i)` means *"compiler-supplied scratch: the hardware writes it, but it carries no result the caller reads"* — and `tile.gather` already carries the exact pair for its own backend-conditional tmp:

```cpp
.add_argument("tmp", "Scratch tile required by TSELS (TileType)")
.set_arg_effect(2, ArgEffect::Write)
.set_workspace_arg(2)
```

MemoryReuse then refuses to let a **written workspace** join a buffer whose other member's lifetime **ends at or before** the workspace's def point.

The guard is **directional**, which keeps the legal opposite case: a later value — typically the TSELS `dst` — may still join the workspace's own buffer, because by then it has retired. `forbid_output_alias(0)` continues to cover `dst` vs `mask` independently.

It is keyed on the registry rather than on operator names, and deliberately **not backend-gated**: PyPTO cannot know which pipe a given intrinsic writes a workspace from, so every declared one is isolated. A5's TSELS happens to leave `tmp` untouched (`vdup` into a vector register); paying a 512-byte buffer there is the price of not encoding per-intrinsic pipe knowledge in the pass. The fact is collected into the shared `AllocationConstraintAnalysis`, so legacy MemoryReuse and DSA-RP both honour it (DSA-RP as a `TargetHazard` separation).

After the fix the workspace takes its own buffer, while the `dst` keeps the legal reuse:

```text
t__ssa_v0          mem_vec_2  @0       sz=8192   pl.tile.load
t__sel_tmp_tmp_v1  mem_vec_7  @16896   sz=512    pl.tile.create   <- own buffer
t__tmp_v2          mem_vec_2  @0       sz=8192   pl.tile.sels     <- reuse preserved
```

`@16896` is the address the workspace already took when the source was kept artificially live, which is the structural confirmation that this is the same placement.

## One existing test changed

`test_sels_output_may_reuse_dead_tmp` reached its `dst == tmp` assertion only *through the hazardous placement*: both landed on `mem_vec_4`, the buffer recycled from a `dead` tile that the immediately preceding `pl.cmps` still reads. The invariant it names — `dst` may alias `tmp`, unlike `mask`/`src` — is real and still worth pinning, so its program now keeps the earlier tiles live and asserts the same thing without relying on the recycled buffer. It passes on both backends.

## Tests

- New `test_sels_scratch_does_not_inherit_a_buffer_a_preceding_op_reads`, parametrized over both backends: A2/A3 must separate the scratch, A5 must keep the tighter packing (pinning that the guard is backend-gated). Fails on `main`.
- The device case `test_composed_selects[select_composed_clamp]` was `xfail` for exactly this bug; the mark is **removed**, so this PR proves itself on hardware in `system-tests-direct`.
- The second `xfail` in that file (A2/A3 INT32 `cmps` mask driving `tsels` inverted) is a different, unfixed defect and stays.
- Full UT suite: 13801 passed. The single failure, `test_symlinked_import_path_still_names_the_caller`, is a pre-existing editable-install issue unrelated to this change.
- clang-format, clang-tidy, cpplint headers, ruff, ruff-format, pyright, markdownlint, and the repo lint scripts are clean.

## Docs

`docs/{en,zh}/dev/passes/36-memory_reuse.md` gain a sibling section next to the existing load + tpop hazard.

---

# Second commit — reject an int32 ordering compare A2/A3 cannot evaluate

Found while explaining the *other* xfail that shipped with `tile.select`. It is an independent defect, but it lands here because it removes the second of the two xfails that PR added, leaving that file with none.

## What

A2/A3 has exactly one int32 comparator, and it is an equality one. Queried through `tbe.common.platform.intrinsic_check_support` — the table TIK's own `tik_params_check.py:622` validates against — on both `Ascend910B1` and `Ascend910_9391`:

```text
              float16  float32    int16   uint16    int32   uint32     int8    uint8
vcmpvs_eq           Y        Y        -        -        Y        -        -        -
vcmpvs_ne           Y        Y        -        -        -        -        -        -
vcmpvs_lt           Y        Y        -        -        -        -        -        -
vcmpvs_le           Y        Y        -        -        -        -        -        -
vcmpvs_gt           Y        Y        -        -        -        -        -        -
vcmpvs_ge           Y        Y        -        -        -        -        -        -
```

pto-isa's `GenCmpCall` short-circuits that one instruction and **discards the requested mode** — `a2a3/TCmps.hpp:48-72` for the scalar form, and `a2a3/TCmp.hpp:24-31` identically for the tile-vs-tile form:

```cpp
if constexpr (std::is_same<TIN, int32_t>::value) {
    vcmpvs_eq(dst, src0, src1, repeat, ...);   // cmpMode never used
} else {
    ... vcmp_dispatch(..., cmpMode, ...)       // every other dtype honours it
}
```

So `lt` / `le` / `gt` / `ge` on an int32 tile silently return an **equality** mask. `TcmpsCheck` explicitly admits `int32_t` in its dtype `static_assert` and never validates `mode`, and PyPTO had no dtype/mode gate either — nothing rejected the pairing at any layer.

Filed upstream as **pto-isa issue #321**.

## The failure it produced

The INT32 case in `test_select.py` (source `-4..4`, `gt`, threshold `0`, false-branch scalar `-3`):

```text
Mismatched elements: 1136/2048
    [4] actual=0,  expected=-3      # v=0: 0 > 0 false -> should take the scalar, kept v
    [5] actual=-3, expected=1       # v=1: 1 > 0 true  -> should keep v, took the scalar
```

Both are exactly what `v == 0` produces. The count matches too — the predicates disagree precisely where `v >= 0`, 5 of the 9 source values, and `2048 * 5/9 ≈ 1138`.

The earlier xfail described this as an "inverted mask" and guessed a mask-granularity mismatch. Both were wrong; it only *reads* as inversion on data where the `gt` and `eq` truth sets happen to be complementary.

## The gate

New `BackendHandler::SupportsOrderingCompareDataType()`, consulted by a shared `MakeCmpCodegenPTO` that now backs both `pto.tcmp` and `pto.tcmps`. The error names the two modes that do work and the workaround:

```text
pto.tcmps cannot evaluate an ordering comparison (cmp_type 2-5: lt/le/gt/ge) on
int32 values for target arch 'a2a3', which has no ordering compare for that
element type and would silently produce an EQUALITY mask instead. Use cmp_type=0
(eq) or cmp_type=1 (ne), or cast the source to FP32 with tile.cast first (exact
for |value| < 2^24).
```

`eq` and `ne` stay available (`ne` is `eq` plus `TCmps`' `vnot` fix-up). A5 is unaffected — its `GenCmpCall` dispatches all six modes generically over `T`.

This is a **denylist of the one combination confirmed wrong on device**, not an allowlist of what is proven right. The 16-bit path bit-casts source and scalar to `half` and borrows the fp16 comparator, which matches integer ordering only for non-negative values; that is analysis rather than a measured failure, so it is reported in issue #321 rather than rejected here.

## Blast radius checked

No existing caller is affected. The only ordering compares in this repo are FP32 (`tests/st/runtime/ops/test_cmp.py`, `test_select.py`), and all 44 in `pypto-lib` (`turboquant_kv.py`, `sample.py`, `lm_head.py`) compare logits / scores / `matmul(out_dtype=FP32)` results.

## Two tests touched

- `test_int32_compare_feeding_a_select` was xfail for exactly this defect. The pairing is now a compile-time error, so the case moves to `cmp_type=ne` — a mode A2/A3 evaluates correctly — keeping real INT32 select coverage on hardware. **`test_select.py` now carries no xfail at all.**
- `test_tsels_emits_mask_src_tmp_and_typed_scalar` used an int32 `gt` only as scaffolding to produce a mask; it now uses `eq` and asserts exactly what it asserted before.

New `test_cmps_rejects_an_int32_ordering_compare_on_a2a3` covers all four ordering modes rejected on A2/A3, both equality modes still emitting, and A5 emitting regardless.

## Verification

Full UT suite: 13802 passed, plus the same single pre-existing editable-install failure. clang-format, clang-tidy, ruff, ruff-format, pyright, markdownlint and the repo lint scripts (including `check_emitter_check_classification`, which governs CHECK vs INTERNAL_CHECK in emitters) are clean.

### CODE REVIEW

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 1]
```diff
@@ -2964,6 +2964,54 @@ class ForbidAliasCollector : public IRVisitor {
...
+ private:
+  /// `tile.sels(mask, src, tmp, scalar)` -- the scratch is operand 2.
+  static constexpr size_t kTselsTmpArgIndex = 2;
+
+  void Record(const OpPtr& op, const std::vector<ExprPtr>& args) {
+    if (!IsOp(op, "tile.sels") || args.size() <= kTselsTmpArgIndex) return;
+    // AsVarLike, not As<Var>: the scratch can reach the op as a loop IterArg
+    // when a select sits inside a loop body.
+    if (auto tmp = AsVarLike(args[kTselsTmpArgIndex])) scratch_.insert(tmp.get());
+  }
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply src/ir/transforms/memory_reuse_pass.cpp]: Confirmed and fixed in 8a5a5c495 — and the gap is wider than an `IterArg`. `interval.variable = sharing_group[0]`, so a lifetime interval is keyed on ONE representative of its sharing group; pointer identity misses a workspace that is not its group's representative, not just one that arrives as an `IterArg` with no interval at all.

Both lookups now resolve through the MemRef allocation base as well as the `Var`, which is the identity the carry chain shares after `MaterializeSemanticAliases` fuses init value / `IterArg` / yield value onto one base. `AllocationHazardInputs` carries a `written_workspace_bases` mirror alongside `written_workspaces`; `is_written_workspace` in the reuse decision and the `TargetHazard` loop in the DSA planner both consult it. The rule stays directional — reuse from an earlier value into the workspace is blocked, a later `dst` joining the dead workspace is not.

Note the design also changed since you reviewed: the `RequiresTselsScratchIsolation()` backend predicate is gone. `tile.sels` now declares `set_arg_effect(2, ArgEffect::Write)` + `set_workspace_arg(2)` — the pair `tile.gather` already carries — and the guard keys on "written workspace" from the registry, with no operator names and no backend gate.

Covered by `test_loop_carried_written_workspace_is_still_isolated`, where the dead value and the carry's init both sit before the loop (the carry's storage is established at its init value, so a value dying inside the body can never be inherited by it — my first attempt at this test put the `cmps` in the body and was vacuous). Verified it fails with the base fallback disabled and passes with it, so it genuinely exercises the path.

**Review by Hzfengsy — APPROVED** [review]
